litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,54 +1922,267 @@ protected LayoutState calculateLayoutState(
+        }
+      });
+
+      this.context = context;
+      this.root = root;
+      this.widthSpec = widthSpec;
+      this.heightSpec = heightSpec;
+      this.diffingEnabled = diffingEnabled;
+      this.previousLayoutState = previousLayoutState;
+      this.treeProps = treeProps;
+      this.extraAttribution = extraAttribution;
+    }
+
+    boolean isReleased() {
+      return isReleased;
+    }
+
+    @Override
+    public void run() {
+      if (runningThreadId.compareAndSet(-1, Process.myTid())) {
+        super.run();
+      }
+    }
+
+    @Override
+    public LayoutState get() {
+      final int runningThreadId = this.runningThreadId.get();
+      final int originalThreadPriority;
+      if (isMainThread() && !isDone() && runningThreadId != Process.myTid()) {
+        // Main thread is about to be blocked, raise the running thread priority.
+        originalThreadPriority = Process.getThreadPriority(runningThreadId);
+
+        boolean success = false;
+        int targetThreadPriority = Process.THREAD_PRIORITY_DISPLAY;
+        while (!success && targetThreadPriority < originalThreadPriority) {
+          // Keep trying to increase thread priority of running thread as long as it is an increase.
+          try {
+            Process.setThreadPriority(runningThreadId, targetThreadPriority);
+            success = true;
+          } catch (SecurityException e) {
+            /*
+              From {@link Process#THREAD_PRIORITY_DISPLAY}, some applications can not change
+              the thread priority to that of the main thread. This catches that potential error
+              and tries to set a lower priority.
+             */
+            targetThreadPriority += Process.THREAD_PRIORITY_LESS_FAVORABLE;
+          }
+        }
+      } else {
+        originalThreadPriority = THREAD_PRIORITY_INVALID;
+      }
+
+      final LayoutState layoutState;
+      try {
+        layoutState = super.get();
+      } catch (ExecutionException e) {
+        final Throwable cause = e.getCause();
+        if (cause instanceof RuntimeException) {
+          throw (RuntimeException) cause;
+        } else {
+          throw new RuntimeException(e.getMessage(), e);
+        }
+      } catch (InterruptedException | CancellationException e) {
+        throw new RuntimeException(e.getMessage(), e);
+      } finally {
+        if (originalThreadPriority != THREAD_PRIORITY_INVALID) {
+          // Reset the running thread's priority after we're unblocked.
+          Process.setThreadPriority(runningThreadId, originalThreadPriority);
+        }
+      }
+
+      // Creating a LayoutState gives it a refcount of 1, but subsequent get() calls to this Future
+      // must increment the refcount to ensure we don't release the LayoutState too early.
+      if (isFirstGet.getAndSet(false)) {
+        return layoutState;
+      }
+      try {
+        return layoutState.acquireRef();
+      } catch (IllegalStateException e) {
+        isReleased = true;
```

Comment:
The isReleased, etc code can be made more explicit: We can have the future keep a ref to the LayoutState and then ref count the future. When the future is released, it releases its ref to the layoutstate as well.

Reply:
This code could probably be replaced with a slightly refactored https://github.com/facebook/litho/blob/master/litho-core/src/main/java/com/facebook/litho/LayoutThreadFactory.java

Reply:
@astreet I think I tried that at first and it doesn't quite work, because most of the code elsewhere references the LayoutState directly and releases that.

If we keep the LayoutState referenced for as long as the Future is referenced, it could lead to "leaking" of the LayoutState longer than expected (e.g. after setRoot to different layout, then setRoot back to original layout), and it risks the referenced LayoutState using older than expected internal state. See my earlier comment when I ran into this: https://github.com/facebook/litho/pull/385#issuecomment-401579714

There was a specific test that was failing, don't remember it right now but I can find out if you're interested.

Reply:
yeah let me make a branch for that other approach and see why it was failing again